tirocks/src/snapshot.rs

```diff
@@ -0,0 +1,171 @@
+// Copyright 2022 TiKV Project Authors. Licensed under Apache-2.0.
+
+use std::{
+    marker::PhantomData,
+    mem::ManuallyDrop,
+    ops::{Deref, DerefMut},
+};
+
+use tirocks_sys::{rocksdb_Iterator, rocksdb_Snapshot};
+
+use crate::{
+    db::{RawCfHandle, RawDbRef},
+    option::ReadOptions,
+    properties::table::user::SequenceNumber,
+    Iterable, PinSlice, RawDb, RawIterator, Result,
+};
+
+pub struct RawSnapshot<'a> {
+    ptr: *const rocksdb_Snapshot,
+    _life: PhantomData<&'a ()>,
+}
+
+impl<'a> RawSnapshot<'a> {
+    #[inline]
+    pub(crate) unsafe fn from_ptr(ptr: *const rocksdb_Snapshot) -> RawSnapshot<'a> {
+        RawSnapshot {
+            ptr,
+            _life: PhantomData,
+        }
+    }
+
+    #[inline]
+    pub(crate) fn as_ptr(&self) -> *const rocksdb_Snapshot {
+        self.ptr
+    }
+
+    #[inline]
+    pub fn sequence_number(&self) -> SequenceNumber {
+        let mut n = 0;
+        unsafe {
+            tirocks_sys::crocksdb_get_snapshot_sequence_number(self.ptr, &mut n);
+        }
+        n
+    }
+}
+
+pub struct WithSnapOpt<'a> {
```

Comment:
Why not just take a copy of `ReadOptions`?

Reply:
It will when necessary. Check https://github.com/BusyJay/tirocks/pull/22/files#diff-487ca0da5dd84b62e9026c612b29409330876daff6c1c8c979c26aefe14ad63dR197-R207.

Reply:
It needs `mut` because it needs to change the snapshot field. And not move it all the time is to make it possible to share read options between ops.

Reply:
I'm not saying the implementation is confusing, just the interface. It shouldn't require user to provide a mutable option.

Reply:
I updated the doc and hopefully it can help understand the design.